src/js/BookReader/Mode1Up.js

```diff
@@ -12,144 +12,219 @@ export class Mode1Up {
-      leafTop += Math.floor(page.height / this.br.reduce) + 10;
+  /**
+   * Find the pages that intersect the current viewport, including 1 before/after
+   **/
+  * findIntersectingPages() {
+    // Rectangle of interest
+    const height = this.$scrollContainer.height();
+    const scrollTop = this.$scrollContainer.scrollTop();
+    const scrollBottom = scrollTop + height;
+    const scrollRegion = { top: scrollTop, bottom: scrollBottom };
+
+    let prev = null;
+    for (const {page, top, bottom} of this.pagesWithBounds()) {
+      const intersects = Mode1Up.boundsIntersect({ top, bottom }, scrollRegion);
+      const cur = {page, top, bottom, intersects};
+      if (intersects) {
+        // Also yield the page just before the visible page
+        if (prev && !prev.intersects) yield prev;
+        yield cur;
+      }
+      // Also yield the page just after the last visible page
+      else if (!cur.intersects && prev?.intersects) {
+        yield cur;
+        break;
+      }
+      prev = cur;
     }
+  }
 
-    for (const index of indicesToDisplay) {
-      const page = book.getPage(index);
-      const height = Math.floor(page.height / this.br.reduce);
-
-      if (!this.br.displayedIndices.includes(index)) {
-        const width = Math.floor(page.width / this.br.reduce);
-        const leftMargin = Math.floor((containerWidth - width) / 2);
-
-        const pageContainer = this.br._createPageContainer(index, {
-          width,
-          height,
-          top: leafTop,
-          left: leftMargin,
-        });
-
-        const pageURISrcset = this.br.options.useSrcSet ? this.br._getPageURISrcset(index, this.br.reduce, 0) : [];
-        const img = $('<img />', {
-          src: page.getURI(this.br.reduce, 0),
-          srcset: pageURISrcset,
-          alt: 'Book page image'
-        });
-        pageContainer.append(img);
-
-        BRpageViewEl.appendChild(pageContainer[0]);
+  drawLeafs() {
+    const pagesToDisplay = Array.from(this.findIntersectingPages());
+
+    if (pagesToDisplay.length) {
+      const documentContainerWidth = this.$documentContainer.width();
+
+      // Based of the pages displayed in the view we set the current index
+      // $$$ we should consider the page in the center of the view to be the current one
+      this.br.updateFirstIndex(pagesToDisplay.find(({intersects}) => intersects).page.index);
+
+      for (const {page, top, bottom} of pagesToDisplay) {
+        if (!this.br.displayedIndices.includes(page.index)) {
+          const height = bottom - top;
+          const width = this.physicalInchesToDisplayPixels(page.widthInches);
+          const reduce = page.width / width;
+
+          this.br._createPageContainer(page.index, {
+            width,
+            height,
+            top,
+            left: Math.floor((documentContainerWidth - width) / 2),
+          })
+            .append($('<img />', {
```

Comment:
q: should we just use the image cache for this, too?

Reply:
Absolutely! In a future PR 😅